### PR TITLE
Imagej live view plugin

### PR DESCRIPTION
**Pull request by ANeaves**
This is a new plugin designed for the [ImageJ software](https://imagej.nih.gov/ij/index.html).

This plugin allows the Live View plugin of Odin Data to stream the images to the ImageJ software. The standard ImageJ tools will then work with the live images. The layout of the plugin's GUI is based on the [Area Detector plugins](http://cars9.uchicago.edu/software/epics/areaDetectorViewers.html).

A build script has been included that can be used to compile the plugin, and create a script called "odin-imagej" in the install path bin directory, which will run ImageJ with the plugin already installed into the plugins menu, and the java variables set to allow zmq to run properly.

### CODE REVIEW

File: tools/imagej/src/LiveViewSocket.java [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+
+    private boolean is_paused = false;
+
+    private Thread zmqThread;
+
+    private static HashMap<String, Integer> dtype_map;
+
+    /**
+     * Constructor for the class. Sets up threads that run the ZMQ socket and background timer which counts the frames.
+     * @param socket_addr    the address of the live view that the socket needs to subscribe to.
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: missing param "parent"

File: tools/imagej/src/LiveViewSocket.java [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+        }
+    }
+
+    /**
+     * Closes down the socket and context, allowing the plugin to close cleanly.
+     * Interrupts the thread running the socket.
+     */
+    public void shutdownSocket()
+    {
+        //printMessage("Socket Shutdown Signal Received. Closing socket");
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: Remove commented out code

File: tools/imagej/src/LiveViewSocket.java [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    }
+
+    /**
+     * Sets up the socket, subscribing it to the supplied socket address
+     * @param addr the socket address to subscribe to.
+     * @throws ZMQException if the socket address supplied is malformed or otherwise invalid.
+     */
+    public void setupSocket(String addr) throws ZMQException
+    {
+        //printMessage("CREATING ZMQ SOCKET");
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: Remove commented out code

File: tools/imagej/src/LiveViewSocket.java [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+        context = ZMQ.context(1);
+        socket = context.socket(ZMQ.SUB);
+        try
+        {
+            socket.connect(addr);
+            this.socket_addr = addr;
+        }
+        catch(ZMQException except)
+        {
+            //printMessage("EXCEPTION CONNECTING SOCKET");
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: Remove commented out code

File: tools/imagej/src/LiveViewSocket.java [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+            throw except;
+            
+        }
+        
+        socket.subscribe("".getBytes());
+        //printMessage(String.format("Subscribed to addr %s",socket_addr));
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: Remove commented out code

File: tools/imagej/src/LiveViewSocket.java [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+                IntBuffer floatBuf = data.asIntBuffer();
+                int[] tmp_array = new int[floatBuf.limit()];
+                img_pixels = new float[floatBuf.limit()];
+                floatBuf.get((int[])tmp_array);
+                for(int i = 0; i < tmp_array.length; i++)
+                {
+                    ((float[])img_pixels)[i] = tmp_array[i];
+                }
+                break;
+
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: Maybe explicitly put in the case for float. Not sure here if it's just been forgotten or whether it's intended for it to go to default.

File: tools/imagej/src/LiveViewSocket.java [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+         * @param dataset  the dataset the image came from. Not always used, depending on the detector
+         * @param bitdepth The bitdepth of the image. Usually either 8, 16, or 32
+         * @param shape    The Dimensions of the data in pixels, as an array of two values: <code>[width, height]</code>
+         */
+        public ImageFrame(String dataset, int bitdepth, int[] shape)
+        {
+            this.dataset = dataset;
+            this.bitdepth = bitdepth;
+            this.shape = shape;
+            this.timestamp = null; //new Date();
```
Comment (MattTaylorDLS) [inline tools/imagej/src/LiveViewSocket.java]: Remove commented out code

File: tools/imagej/src/Live_View.java [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+
+	long last_image_time = new Date().getTime();
+	Queue<Long> time_avg_queue;
+	long total = 0;
+	int avg_size = 10; //size of the queue used for the rolling average of time between frames
+
+	boolean is_logging = false;
+
+	private TimerTask status_update;
+    private Timer frame_counter;
```
Comment (MattTaylorDLS) [inline tools/imagej/src/Live_View.java]: indentation

File: tools/imagej/src/Live_View.java [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+				instance = null;
+				frame_counter.cancel();
+			}
+		});
+
+		Runtime.getRuntime().addShutdownHook(new Thread()
+		{
+			@Override
+			public void run(){
+				// printMessage("Interrupt received, shutting down plugin");
```
Comment (MattTaylorDLS) [inline tools/imagej/src/Live_View.java]: Remove commented out code

File: tools/imagej/src/Live_View.java [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+	private TimerTask status_update;
+    private Timer frame_counter;
+
+	/**
+	 * Constructs the Live View Plugin.
+	 * It creates and displays the Control GUI and sets up required objects.
+	 * It also sets up the action listeners that listen for the window closing or being shut down, so it
+	 * can tear down cleanly.
+	 */
+	public Live_View()
```
Comment (MattTaylorDLS) [inline tools/imagej/src/Live_View.java]: Seems to be an odd mix of indentation in this class compared to others. Tabs and spaces

File: tools/imagej/src/Live_View.java [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+		btn_live_on_off.addActionListener(new ActionListener()
+		{
+			public void actionPerformed(ActionEvent event)
+			{
+				if(socket != null){
+					boolean is_paused = socket.invertPaused();		
+					if(!is_paused)
+					{
+						btn_live_on_off.setLabel("PAUSE");
+						// txt_socket_addr.setBackground(Color.green);
```
Comment (MattTaylorDLS) [inline tools/imagej/src/Live_View.java]: Remove commented out code

File: tools/imagej/src/Live_View.java [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+					}
+					else
+					{
+						btn_live_on_off.setLabel("UNPAUSE");
+						// txt_socket_addr.setBackground(Color.red);
```
Comment (MattTaylorDLS) [inline tools/imagej/src/Live_View.java]: Remove commented out code

File: tools/imagej/src/Live_View.java [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+
+	/**
+	 * Creates the Live View socket and collects the required references from the newly created socket.
+	 * If the socket cannot connect, the method will fail and a IJ Error window displayed.
+	 */
+	private void activateSocket()
+	{
+		try
+		{
+		socket = new LiveViewSocket(socket_addr, this);
```
Comment (MattTaylorDLS) [inline tools/imagej/src/Live_View.java]: indentation

File: tools/python/test/static/js/live_view.js [hunk 14]
```diff
@@ -65,17 +95,18 @@ var LiveViewApp = (function()
             }
         });
 
+        updateImage();
         // Start the update polling loop
-        pollUpdate();
+        //pollUpdate();
```
Comment (MattTaylorDLS) [inline tools/python/test/static/js/live_view.js]: Remove commented out code

File: tools/python/test/static/js/live_view.js [hunk 15]
```diff
@@ -65,17 +95,18 @@ var LiveViewApp = (function()
...
-        }
-        setTimeout(pollUpdate, 100);      
-    };
+//    var pollUpdate = function()
+//    {
+//        if (liveview_enable) {
+//            updateImage();
+//        }
+//        setTimeout(pollUpdate, 100);
+//    };
```
Comment (MattTaylorDLS) [inline tools/python/test/static/js/live_view.js]: Remove commented out code

File: tools/imagej/src/LiveViewSocket.java (hunk 1 above)
Comment (ANeaves) [reply tools/imagej/src/LiveViewSocket.java]: Added parent to javadoc

File: tools/imagej/src/Live_View.java (hunk 8 above)
Comment (ANeaves) [reply tools/imagej/src/Live_View.java]: Not sure what caused this mess, but all tabs have now been replaced by 4 spaces to be consistent.

**Comment by ajgdls** [comment]
Hi @ANeaves can you give me some guidance regarding json.jar.  I'm getting an error message stating that I need to download json.jar.  Is this a standard java archive or from a project?  I think maybe a README which has come installation instructions could be useful as there are quite a few dependencies for this plugin.  I've git cloned and built jzmq which I assume is correct.  Thanks.

**Comment by ajgdls** [comment]
Hi @ANeaves forget the last comment I've just seen the install README.

**Comment by ajgdls** [comment]
Hello again @ANeaves I am having some trouble with the execution of the script.  I can now run the build script which appears to complete.  The odin-imagej startup script is created but it does not have the correct permissions for execution.  I manually changed these and executed and it ran OK.  However when I tried to click on the connect button of the plugin I got the following error:

**Comment by ajgdls** [comment]
Hi @ANeaves no they are installed into different locations.  I installed JZMQ into a prefix directory within its own structure and then the imagej prefix is within the odin-data structure.  I believe I correctly set the flag to point to JZMQ during the build (but I could have got it wrong but the build did complete OK).
The way the Diamond build system works for tools means that we would need to install each tool/library into its own area.
Thanks for looking into it.

**Comment by ANeaves** [comment]
Hey @ajgdls, I've managed to replicate your error message. It looks like you did not set the `--with-zmq-library` flag correctly and it defaulted to pointing at `$INSTALL_PREFIX/lib` as the directory containing the libzmq stuff. The build script completed OK because it only checks if that directory exists, not if it contains the files needed, a mistake I've now fixed. If you try and run the script again it should not pass without finding *libjzmq.so* in the directory provided by `--with-zmq-library`

**Comment by ajgdls** [comment]
Great thank you @ANeaves it now works for me with your build script.

**Comment by ANeaves** [comment]
hey @GDYendell @ulrikpedersen Any chance of a quick look today/tomorrow? I would like to get this merged ASAP, been a long time coming!

**Pull request merged**